plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -356,6 +377,160 @@ private static ConnectorPageSource createOrcPageSource(
         }
     }
 
+    private static List<Integer> applyProjection(ColumnHandle expectedColumnHandle, ColumnHandle readColumnHandle)
+    {
+        IcebergColumnHandle expectedColumn = (IcebergColumnHandle) expectedColumnHandle;
+        IcebergColumnHandle readColumn = (IcebergColumnHandle) readColumnHandle;
+        checkState(readColumn.isParentOf(expectedColumn), "Read column path must be a prefix of the Projected column path");
+
+        List<String> expectedDereferences = expectedColumn.getFullPath();
+        List<String> readerDereferences = readColumn.getFullPath();
+
+        List<String> path = expectedDereferences.subList(readerDereferences.size(), expectedDereferences.size());
+        ImmutableList.Builder<Integer> dereferenceChain = ImmutableList.builder();
+        ColumnIdentity columnIdentity = readColumn.getColumnIdentity();
+        for (String dereference : path) {
+            List<ColumnIdentity> children = columnIdentity.getChildren();
+            for (int i = 0; i < children.size(); i++) {
+                if (children.get(i).getName().equals(dereference)) {
+                    dereferenceChain.add(i);
+                    columnIdentity = children.get(i);
+                    break;
+                }
+            }
+        }
+
+        return dereferenceChain.build();
+    }
+
+    /**
+     * Creates a mapping between the input {@param columns} and base columns if required.
+     */
+    public static Optional<ReaderColumns> projectColumnsForOrc(List<IcebergColumnHandle> columns)
+    {
+        // No projection is required if all columns are base columns
+        if (columns.stream().allMatch(IcebergColumnHandle::isBaseColumn)) {
+            return Optional.empty();
+        }
+
+        Map<IcebergColumnHandle, Integer> originalColumnIndexes = Streams.mapWithIndex(columns.stream(), Map::entry)
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> toIntExact(entry.getValue())));
+        // Order the columns in path depth order so existing projections for nested columns can be reused
+        List<IcebergColumnHandle> orderedColumns = originalColumnIndexes.keySet().stream()
+                .sorted(Comparator.comparingInt(col -> col.getPath().size()))
+                .collect(toImmutableList());
+
+        ImmutableList.Builder<ColumnHandle> projectedColumns = ImmutableList.builder();
+        Integer[] outputColumnMapping = new Integer[columns.size()];
+        Map<List<String>, Integer> mappedColumnIndices = new HashMap<>();
+        int projectedColumnIndex = 0;
+
+        for (IcebergColumnHandle column : orderedColumns) {
+            Optional<Integer> existingProjectedColumn = findExistingProjection(mappedColumnIndices, column);
```

Comment:
Please ensure that we have a testcase when we are projecting two fields when one is child of the other (x.y and x.y.z). Maybe it is there - I did not check.

Reply:
Pretty sure this is what you're talking about: https://github.com/trinodb/trino/pull/8129/files#diff-8b18362088e5955f03b860a52d3ee8d3002f053c9e5ca76a888c27a272f9f0b7R1925

That one is effectively `select w, w.x.y.z`, nested a few levels deeper